File: shared/node/socq-client.mjs
const TERMINAL_STATES = new Set(["succeeded", "failed"]);

function sleep(milliseconds) {
  return new Promise((resolve) => setTimeout(resolve, milliseconds));
}

export class SocQClient {
  constructor({
    apiKey = process.env.SOCQ_API_KEY,
    baseUrl = process.env.SOCQ_BASE_URL || "https://api.socq.ai",
  } = {}) {
    if (!apiKey || apiKey === "your-api-key") {
      throw new Error("Set SOCQ_API_KEY before running this example.");
    }

    this.apiKey = apiKey;
    this.baseUrl = baseUrl.replace(/\/+$/, "");
  }

  async request(path, init = {}) {
    const response = await fetch(`${this.baseUrl}${path}`, {
      ...init,
      headers: {
        Authorization: `Bearer ${this.apiKey}`,
        "Content-Type": "application/json",
        ...init.headers,
      },
    });

    const text = await response.text();
    const body = text ? JSON.parse(text) : null;

    if (!response.ok) {
      const message = body?.error?.message || body?.message || `HTTP ${response.status}`;
      throw new Error(`SocQ request failed: ${message}`);
    }

    return body;
  }

  submit(path, payload) {
    return this.request(path, {
      method: "POST",
      body: JSON.stringify(payload),
    });
  }

  getTask(taskId, { cursor, limit = 50 } = {}) {
    const query = new URLSearchParams({ limit: String(limit) });
    if (cursor) query.set("cursor", cursor);
    return this.request(`/v1/tasks/${encodeURIComponent(taskId)}?${query}`);
  }

  async waitForTask(taskId, { intervalMs = 2000, timeoutMs = 120000 } = {}) {
    const deadline = Date.now() + timeoutMs;

    while (Date.now() < deadline) {
      const response = await this.getTask(taskId);
      const task = response?.data;

      if (!task?.status) {
        throw new Error("SocQ task response did not include data.status.");
      }

      if (TERMINAL_STATES.has(task.status)) {
        if (task.status === "failed") {
          throw new Error(task.error_message || `SocQ task ${taskId} failed.`);
        }
        return task;
      }

      await sleep(intervalMs);
    }

    throw new Error(`Timed out waiting for SocQ task ${taskId}.`);
  }
}

File: shared/python/socq_client.py
import os
import time
from typing import Any, Dict, Optional
from urllib.parse import quote

import requests


TERMINAL_STATES = {"succeeded", "failed"}


class SocQClient:
    def __init__(
        self,
        api_key: Optional[str] = None,
        base_url: Optional[str] = None,
    ) -> None:
        self.api_key = api_key or os.environ.get("SOCQ_API_KEY")
        self.base_url = (base_url or os.environ.get("SOCQ_BASE_URL") or "https://api.socq.ai").rstrip("/")

        if not self.api_key or self.api_key == "your-api-key":
            raise ValueError("Set SOCQ_API_KEY before running this example.")

    def request(self, method: str, path: str, **kwargs: Any) -> Dict[str, Any]:
        headers = {
            "Authorization": "Bearer {}".format(self.api_key),
            "Content-Type": "application/json",
        }
        response = requests.request(
            method,
            "{}{}".format(self.base_url, path),
            headers=headers,
            timeout=30,
            **kwargs,
        )

        try:
            body = response.json()
        except ValueError:
            body = None

        if not response.ok:
            message = "HTTP {}".format(response.status_code)
            if isinstance(body, dict):
                error = body.get("error")
                if isinstance(error, dict):
                    message = error.get("message") or message
                else:
                    message = body.get("message") or message
            raise RuntimeError("SocQ request failed: {}".format(message))

        if not isinstance(body, dict):
            raise RuntimeError("SocQ returned an unexpected response.")
        return body

    def submit(self, path: str, payload: Dict[str, Any]) -> Dict[str, Any]:
        return self.request("POST", path, json=payload)

    def get_task(
        self,
        task_id: str,
        cursor: Optional[str] = None,
        limit: int = 50,
    ) -> Dict[str, Any]:
        params: Dict[str, Any] = {"limit": limit}
        if cursor:
            params["cursor"] = cursor
        return self.request("GET", "/v1/tasks/{}".format(quote(task_id, safe="")), params=params)

    def wait_for_task(
        self,
        task_id: str,
        interval_seconds: float = 2,
        timeout_seconds: float = 120,
    ) -> Dict[str, Any]:
        deadline = time.time() + timeout_seconds

        while time.time() < deadline:
            response = self.get_task(task_id)
            task = response.get("data") or {}
            status = task.get("status")

            if not status:
                raise RuntimeError("SocQ task response did not include data.status.")
            if status in TERMINAL_STATES:
                if status == "failed":
                    raise RuntimeError(task.get("error_message") or "SocQ task failed.")
                return task

            time.sleep(interval_seconds)

        raise TimeoutError("Timed out waiting for SocQ task {}.".format(task_id))
